lib/model/model.go

```diff
@@ -1151,46 +1278,94 @@ func (m *model) ClusterConfig(deviceID protocol.DeviceID, cm protocol.ClusterCon
+	hasTokenRemote := hasDeviceRemote && len(ccDeviceRemote.EncPwToken) > 0
+	hasTokenLocal := hasDeviceLocal && len(ccDeviceLocal.EncPwToken) > 0
+	isEncRemote := folderDevice.EncryptionPassword != ""
+	isEncLocal := fcfg.Type == config.FolderTypeReceiveEncrypted
+
+	if !isEncRemote && !isEncLocal && deviceUntrusted {
+		return errEncNotEncryptedUntrusted
 	}
 
-	if deviceCfg.Introducer {
-		folders, devices, foldersDevices, introduced := m.handleIntroductions(deviceCfg, cm)
-		folders, devices, deintroduced := m.handleDeintroductions(deviceCfg, foldersDevices, folders, devices)
-		if introduced || deintroduced {
-			changed = true
-			cfg := m.cfg.RawCopy()
-			cfg.Folders = make([]config.FolderConfiguration, 0, len(folders))
-			for _, fcfg := range folders {
-				cfg.Folders = append(cfg.Folders, fcfg)
-			}
-			cfg.Devices = make([]config.DeviceConfiguration, len(devices))
-			for _, dcfg := range devices {
-				cfg.Devices = append(cfg.Devices, dcfg)
-			}
-			m.cfg.Replace(cfg)
-		}
+	if !(hasTokenRemote || hasTokenLocal || isEncRemote || isEncLocal) {
+		// Noone cares about encryption here
+		return nil
 	}
 
-	if changed {
-		if err := m.cfg.Save(); err != nil {
-			l.Warnln("Failed to save config", err)
+	if isEncRemote && isEncLocal {
+		// Should never happen, but config racyness and be safe.
+		return errEncInvConfigLocal
+	}
+
+	if hasTokenRemote && hasTokenLocal {
+		return errEncInvConfigRemote
+	}
+
+	if !(hasTokenRemote || hasTokenLocal) {
+		return errEncNotEncryptedRemote
+	}
+
+	if !(isEncRemote || isEncLocal) {
+		return errEncNotEncryptedLocal
+	}
+
+	if isEncRemote {
+		pwToken := protocol.PasswordToken(fcfg.ID, folderDevice.EncryptionPassword)
+		match := false
+		if hasTokenLocal {
+			match = bytes.Equal(pwToken, ccDeviceLocal.EncPwToken)
+		} else {
+			// hasTokenRemote == true
+			match = bytes.Equal(pwToken, ccDeviceRemote.EncPwToken)
+		}
+		if !match {
+			return errEncPW
 		}
+		return nil
 	}
 
+	// isEncLocal == true
+
+	var ccToken []byte
+	if hasTokenLocal {
+		ccToken = ccDeviceLocal.EncPwToken
+	} else {
+		// hasTokenRemote == true
+		ccToken = ccDeviceRemote.EncPwToken
+	}
+	m.fmut.RLock()
+	token, ok := m.folderEncPwTokens[fcfg.ID]
+	m.fmut.RUnlock()
+	if !ok {
+		var err error
+		token, err = readEncToken(fcfg)
+		if err != nil && !fs.IsNotExist(err) {
+			return err
+		}
+		if err == nil {
+			m.fmut.Lock()
+			m.folderEncPwTokens[fcfg.ID] = token
+			m.fmut.Unlock()
+		} else {
+			if err := writeEncToken(ccToken, fcfg); err != nil {
+				return err
+			}
+			m.fmut.Lock()
+			m.folderEncPwTokens[fcfg.ID] = ccToken
+			m.fmut.Unlock()
+			// We can only announce ourselfs once we have the token,
+			// thus we need to resend CCs now that we have it.
+			m.closeConns(fcfg.DeviceIDs(), errEncReceivedToken)
+			return errEncReceivedToken
```

Comment:
Sorry, but I have no idea what this is. I can't seem to pin it to anything that was previously discussed.

Reply:
When a receive-encrypted folder is first created, it couldn't have received the password token yet. Only once it receives the first cluster config from a remote, it does receive the token. However it must send that token for a connection to be successfully established, as the remote needs to be able to check that it has the right password.  
Here we just received that first token, thus close the connection to resend the cluster config with that token included.